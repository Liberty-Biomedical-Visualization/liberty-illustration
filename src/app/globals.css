@tailwind base;
@tailwind components;
@tailwind utilities;

:root {
  --accent-50: 224 247 252;
  --accent-100: 178 224 240;
  --accent-200: 128 200 232;
  --accent-300: 77 176 215;
  --accent-400: 26 153 206;
  --accent-500: 3 152 206;
  --accent-600: 2 128 165;
  --accent-700: 2 109 136;
  --accent-800: 2 90 107;
  --accent-900: 2 61 77;
  --accent-950: 1 44 58;
  --background: 255 255 255;
  --foreground: 23 23 23;
}

body {
  color: var(--foreground);
  background: var(--background);
  font-family: Arial, Helvetica, sans-serif;
}

@layer utilities {
  .text-balance {
    text-wrap: balance;
  }
}
