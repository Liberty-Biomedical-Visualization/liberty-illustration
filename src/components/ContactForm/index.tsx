"use client";

import { useRouter } from "next/navigation";
import { FormEvent, useId, useState } from "react";

import type { PostContactBody } from "@/app/api/contact/route";
import type postContact from "@/lib/api/postContact";
import resolveClassNames from "@/lib/resolveClassNames";

import Input from "../Input";
import withPostContact from "../withPostContact";

export default function ContactForm(props: Readonly<ContactFormProps>) {
  const [email, setEmail] = useState("");
  const [isInErrorState, setIsInErrorState] = useState(false);
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [message, setMessage] = useState("");
  const [subject, setSubject] = useState("");
  const [validation, setValidation] = useState<Partial<PostContactBody> | null>(
    null,
  );
  const router = useRouter();

  const handleSubmit = async (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();

    setIsInErrorState(false);
    setIsSubmitting(true);
    setValidation(null);

    const body = { email, subject, message };
    const result = await props.postContact(body, fetch);

    if (result.isOk) {
      return router.push("/contact/success");
    }

    if (result.validation) {
      setValidation(result.validation);
    } else {
      setIsInErrorState(true);
    }

    setIsSubmitting(false);
  };

  const errorId = useId();

  const className = resolveClassNames(
    "gap-x-4 grid grid-cols-1 max-w-prose mx-auto sm:grid-cols-2",
    props.className,
  );

  return (
    <form
      aria-describedby={isInErrorState ? errorId : undefined}
      className={className}
      onSubmit={handleSubmit}
    >
      <Input
        disabled={isSubmitting}
        label="Email"
        name="email"
        onChange={(event) => setEmail(event.target.value)}
        type="email"
        validation={validation?.email ?? null}
        value={email}
      />
      <Input
        disabled={isSubmitting}
        label="Subject"
        name="subject"
        onChange={(event) => setSubject(event.target.value)}
        type="text"
        validation={validation?.subject ?? null}
        value={subject}
      />
      <Input
        className="sm:col-span-2"
        disabled={isSubmitting}
        label="Message"
        name="message"
        onChange={(event) => setMessage(event.target.value)}
        type="textarea"
        validation={validation?.message ?? null}
        value={message}
      />
      <button
        className="active:bg-accent-200 bg-accent-500 col-span-1 sm:col-start-2 disabled:bg-accent-700 hover:bg-accent-300 mt-4 p-2 text-background text-lg transition-colors"
        disabled={isSubmitting}
        type="submit"
      >
        Submit
      </button>
      {isInErrorState && (
        <div
          className="max-w-prose mt-4 sm:col-span-2 text-red-600"
          id={errorId}
          role="alert"
        >
          <strong>Oops! Something went wrong.</strong>
          <p>
            We’re experiencing an issue on our side and couldn’t process your
            request. Please try again. If the problem persists, please try again
            later.
          </p>
        </div>
      )}
    </form>
  );
}

export interface ContactFormProps {
  className?: string;
  postContact: typeof postContact;
}

export const ContactFormWithPostContact = withPostContact(ContactForm);
