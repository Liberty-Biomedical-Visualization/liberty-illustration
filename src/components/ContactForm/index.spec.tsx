import { describe, expect, it, jest } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import {
  AppRouterContext,
  type AppRouterInstance,
} from "next/dist/shared/lib/app-router-context.shared-runtime";

import ContactForm, { type ContactFormProps } from ".";

describe("ContactForm", () => {
  // fetch is not globally available in the test environment, but it is passed
  // to postContact as a dependency.
  global.fetch = () => Promise.resolve(new Response());

  it("should display an Email field", () => {
    renderContactForm();
    const result = screen.getByLabelText("Email");
    expect(result).toBeInTheDocument();
  });

  it("should display a Subject field", () => {
    renderContactForm();
    const result = screen.getByLabelText("Subject");
    expect(result).toBeInTheDocument();
  });

  it("should display a Message field", () => {
    renderContactForm();
    const result = screen.getByLabelText("Message");
    expect(result).toBeInTheDocument();
  });

  it("should call postContact with field values when form is submitted", async () => {
    const email = "foo@bar.com";
    const subject = "Test Subject";
    const message = "Test message, please ignore.";
    const postContact = jest.fn(() => Promise.resolve({ isOk: true }));
    renderContactForm({ postContact });

    const user = userEvent.setup();
    const emailField = screen.getByLabelText("Email");
    await user.type(emailField, email);
    const subjectField = screen.getByLabelText("Subject");
    await user.type(subjectField, subject);
    const messageField = screen.getByLabelText("Message");
    await user.type(messageField, message);
    const submitButton = screen.getByRole("button", { name: "Submit" });
    await user.click(submitButton);

    expect(postContact).toHaveBeenCalledWith(
      { email, subject, message },
      fetch,
    );
  });

  it("should redirect to the contact success page when postContact result is OK", async () => {
    const router = createMockRouter();
    const postContact = () => Promise.resolve({ isOk: true });
    renderContactForm({ postContact }, router);

    await fillAndSubmitContactForm();

    expect(router.push).toHaveBeenCalledWith("/contact/success");
  });

  it("should display validation errors when postContact result contains validation", async () => {
    const emailValidation = "Ridiculous";
    const messageValidation = "Silly";
    const subjectValidaiton = "Outrageous";
    const postContact = () =>
      Promise.resolve({
        isOk: false,
        validation: {
          email: emailValidation,
          message: messageValidation,
          subject: subjectValidaiton,
        },
      });
    renderContactForm({ postContact });

    await fillAndSubmitContactForm();

    const emailField = screen.getByLabelText("Email");
    expect(emailField).toHaveAccessibleDescription(emailValidation);
    const messageField = screen.getByLabelText("Message");
    expect(messageField).toHaveAccessibleDescription(messageValidation);
    const subjectField = screen.getByLabelText("Subject");
    expect(subjectField).toHaveAccessibleDescription(subjectValidaiton);
  });

  it("should not redirect when postContact result is not OK", async () => {
    const router = createMockRouter();
    const postContact = () => Promise.resolve({ isOk: false });
    renderContactForm({ postContact }, router);

    await fillAndSubmitContactForm();

    expect(router.push).not.toHaveBeenCalled();
  });

  it("should display an error message when postContact result is not OK and does not contain validation", async () => {
    const postContact = () => Promise.resolve({ isOk: false });
    renderContactForm({ postContact });

    await fillAndSubmitContactForm();

    const alert = screen.getByRole("alert");
    expect(alert).toHaveTextContent("Oops! Something went wrong.");
  });

  it("should not redirect when the server returns a 500 response", () => {});

  it("should match the snapshot", () => {
    const result = renderContactForm();
    expect(result.container).toMatchSnapshot();
  });
});

function createMockRouter(): AppRouterInstance {
  return {
    back: jest.fn(),
    forward: jest.fn(),
    prefetch: jest.fn(),
    push: jest.fn(),
    refresh: jest.fn(),
    replace: jest.fn(),
  };
}

async function fillAndSubmitContactForm() {
  const user = userEvent.setup();
  const emailField = screen.getByLabelText("Email");
  await user.type(emailField, "foo@bar.com");
  const subjectField = screen.getByLabelText("Subject");
  await user.type(subjectField, "Test Subject");
  const messageField = screen.getByLabelText("Message");
  await user.type(messageField, "Test message, please ignore.");
  const submitButton = screen.getByRole("button", { name: "Submit" });
  await user.click(submitButton);
}

function renderContactForm(
  props?: Partial<Readonly<ContactFormProps>>,
  router: AppRouterInstance = createMockRouter(),
) {
  const resolvedProps = {
    postContact: () => Promise.resolve({ isOk: true }),
    ...props,
  };

  return render(
    <AppRouterContext.Provider value={router}>
      <ContactForm {...resolvedProps} />
    </AppRouterContext.Provider>,
  );
}
