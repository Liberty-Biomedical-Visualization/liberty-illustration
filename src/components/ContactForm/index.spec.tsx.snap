// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm should match the snapshot 1`] = `
<div>
  <form
    class="gap-x-4 grid grid-cols-1 max-w-prose mx-auto sm:grid-cols-2"
  >
    <div
      class="flex flex-col pb-4 relative"
    >
      <label
        for=":r11:"
      >
        Email
      </label>
      <input
        class="border p-2"
        id=":r11:"
        name="email"
        required=""
        type="email"
        value=""
      />
    </div>
    <div
      class="flex flex-col pb-4 relative"
    >
      <label
        for=":r12:"
      >
        Subject
      </label>
      <input
        class="border p-2"
        id=":r12:"
        name="subject"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="flex flex-col pb-4 relative sm:col-span-2"
    >
      <label
        for=":r13:"
      >
        Message
      </label>
      <textarea
        class="border p-2"
        id=":r13:"
        name="message"
        required=""
      />
    </div>
    <button
      class="active:bg-accent-200 bg-accent-500 col-span-1 sm:col-start-2 disabled:bg-accent-700 hover:bg-accent-300 mt-4 p-2 text-background text-lg transition-colors"
      type="submit"
    >
      Submit
    </button>
  </form>
</div>
`;
